eras/alonzo/formal-spec/mir.tex

```diff
@@ -0,0 +1,415 @@
+\section{MIR Certificates}
+\label{sec:mir}
+
+\newcommand{\InstantaneousRewards}{\type{InstantaneousRewards}}
+\newcommand{\MIRPot}{\type{MIRPot}}
+\newcommand{\ReservesMIR}{\type{ReservesMIR}}
+\newcommand{\TreasuryMIR}{\type{TreasuryMIR}}
+
+In the Alonzo era, MIR certs have:
+
+\begin{itemize}
+  \item Slightly different behavior -
+  prior to the Alonzo era, MIR certificates within the same epoch override
+  values for repeated stake addresses.
+  In the Alonzo era, repetitions are handled by adding the values.
+  Moreover, negative values are allowed, so long as the change never results in
+  an overall negative sum.
+  See Figure \ref{fig:dcert-mir}.
+  \item Extended functionality -
+  MIR certificates can now be used to transfer funds between the reserves and the treasury.
+  See Figure \ref{fig:dcert-mir-trans}.
+\end{itemize}
+
+Figure \ref{fig:defs:alonzo-mir} describes the necessary changes to the types.
+Note that $\fun{mirTarget}$ was named $\fun{credCoinMap}$ in the Shelley specification.
+Figure \ref{fig:dcert-mir} is a modification to the DELEG rule, and
+Figure \ref{fig:dcert-mir-trans} is an addition to the DELEG rule.
+Lastly, Figure \ref{fig:rules:mir} is the updated MIR rule.
+
+
+\begin{figure*}[htb]
+  \emph{Derived types}
+  %
+  \begin{equation*}
+    \begin{array}{llr}
+      \InstantaneousRewards =
+      & (\StakeCredential \mapsto \Coin)
+      & \text{rewards from reserves}
+      \\
+      & \times (\StakeCredential \mapsto \Coin)
+      & \text{rewards from treasury}
+      \\
+      & \times \hldiff{\Coin}
+      & \text{pending modification to reserves}
+      \\
+      & \times \hldiff{\Coin}
+      & \text{pending modification to treasury}
+    \end{array}
+  \end{equation*}
+  %
+  \emph{New MIR Cert accessor}
+  %
+  \begin{equation*}
+    \begin{array}{r@{~:~}lr}
+      \fun{mirTarget}
+      & \DCertMir \to (\StakeCredential \mapsto \Coin) \uniondistinct \hldiff{\MIRPot}
+    \end{array}
+  \end{equation*}
+  \caption{Alonzo Types for MIR updates}
+  \label{fig:defs:alonzo-mir}
+\end{figure*}
+
+\begin{figure}[htp]
```

Comment:
There's a lot of duplication in this and the next figure. I see two options of simplifying them:
1. Define `i'_rwd` depending on the two cases for `mirPot c` and drop the condition. This would give us two rules with still some duplication, but a lot less than now.
2. Make a new STS that just transitions the state of `i_rwd`. Then there would only be a single rule for `DELEG` that handles all the common stuff and the new STS handles all the special cases. This would be more complicated but not have any duplication. Probably not worth it, except if we expect more changes in the future.

Reply:
yea, #1 as actually how the code works, but I wasn't sure how to make it readable in the spec.

https://github.com/input-output-hk/cardano-ledger/blob/fdcb83f35f58d0320f8a0bcf6c1ae54395bb984c/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs#L328-L335

Reply:
option 2 would be too different from the implementation for my comfort...

Let me think about how to do 1 nicely in latex w/o programming jargon :thinking:

Reply:
Let me recommend the `cases` environment :)

Reply:
in a let binding? or use helper functions?

Reply:
ok @WhatisRT , I've tried out using cases, let me know what you think. it's nice that it actually matches the code a lot better now.